lib/cusolver/cusolverRF.jl

```diff
@@ -0,0 +1,436 @@
+
+    # LU Factorization
+    info = Ref{CUSOLVER.csrqrInfo_t}()
+    cusolverSpCreateCsrluInfoHost(info)
+
+    cusolverSpXcsrluAnalysisHost(
+        spH,
+        m, nnzA, desca,
+        h_rowsB, h_colsB, info[],
+    )
+
+    size_internal = Ref{Cint}(0)
+    size_lu = Ref{Cint}(0)
+    cusolverSpDcsrluBufferInfoHost(
+        spH,
+        n, nnzA, desca,
+        h_valsB, h_rowsB, h_colsB,
+        info[],
+        size_internal, size_lu
+    )
+
+    n_bytes = size_lu[] * sizeof(Cint)
+    buffer_lu = zeros(Cint, size_lu[])
+    pivot_threshold = 1.0
+
+    cusolverSpDcsrluFactorHost(
+        spH, n, nnzA, desca,
+        h_valsB, h_rowsB, h_colsB,
+        info[], pivot_threshold,
+        buffer_lu,
+    )
+
+    # Check singularity
+    if check
+        singularity = Ref{Cint}(0)
+        cusolverSpDcsrluZeroPivotHost(
+            spH, info[], tol, singularity,
+        )
+
+        # Check that the matrix is nonsingular
+        if singularity[] >= 0
+            SingularException(singularity[])
+        end
+    end
+
+    # Get size of L and U
+    pnnzU = Ref{Cint}(0)
+    pnnzL = Ref{Cint}(0)
+    cusolverSpXcsrluNnzHost(
+        spH,
+        pnnzL, pnnzU, info[],
+    )
+
+    nnzL = pnnzL[]
+    nnzU = pnnzU[]
+
+    # Retrieve L and U matrices
+    h_Plu = zeros(Cint, m)
+    h_Qlu = zeros(Cint, n)
+
+    h_valsL = zeros(nnzL)
+    h_rowsL = zeros(Cint, m+1)
+    h_colsL = zeros(Cint, nnzL)
+
+    h_valsU = zeros(nnzU)
+    h_rowsU = zeros(Cint, m+1)
+    h_colsU = zeros(Cint, nnzU)
+
+    # Extract
+    cusolverSpDcsrluExtractHost(
+        spH,
+        h_Plu, h_Qlu,
+        desca,
+        h_valsL, h_rowsL, h_colsL,
+        desca,
+        h_valsU, h_rowsU, h_colsU,
+        info[],
+        buffer_lu,
+    )
+
+    h_P = h_Qreorder[h_Plu .+ 1]
+    h_Q = h_Qreorder[h_Qlu .+ 1]
+
+    return RfHostLU(
+        nnzA, h_rowsA, h_colsA, h_valsA,
+        nnzL, h_rowsL, h_colsL, h_valsL,
+        nnzU, h_rowsU, h_colsU, h_valsU,
+        h_P, h_Q,
+    )
+end
+
+struct RfLU{T} <: LinearAlgebra.Factorization{T}
+    rf::RfHandle
+    nrhs::Int
+    n::Int
+    m::Int
+    nnzA::Int
+    drowsA::CuVector{Cint}
+    dcolsA::CuVector{Cint}
+    dP::CuVector{Cint}
+    dQ::CuVector{Cint}
+    dT::CuVector{T}
+end
+
+function RfLU(
+    A::CuSparseMatrixCSR{T, Ti};
+    nrhs=1, ordering=:AMD, check=true, fast_mode=true,
+    factorization_algo=CUSOLVERRF_FACTORIZATION_ALG0,
+    triangular_algo=CUSOLVERRF_TRIANGULAR_SOLVE_ALG1,
+) where {T, Ti}
+    if nrhs > 1
+        error("Currently CusolverRF supports only one right-hand side.")
+    end
+    n, m = size(A)
+    lu_host = RfHostLU(A; ordering=ordering, check=check)
+
+    # Allocations (device)
+    d_T = CUDA.zeros(Cdouble, m * nrhs)
+
+    rf = sparse_rf_handle(;
+        fast_mode=fast_mode,
+        factorization_algo=factorization_algo,
+        triangular_algo=triangular_algo,
+    )
+
+    # Assemble internal data structures
+    cusolverRfSetupHost(
```

Comment:
I confess the current API is misleading. In `cusolverRf`, we initiate a new instance `rf` with a LU factorization we computed before (on the host or on the device). Here, we use cusolver's internal functions to compute the initial LU factorization on the host (the functions are exposed in the header `cusolver.h`, but are not documented). The functions used when we call `RfHostLU` are exactly the same one used when we call `csrlsvlu!` in `CUDA.CUSOLVER`: `cusolver` takes as input a `CuSparseMatrixCSR`, but moves afterwards all the data on the host to compute the LU factorization. 

As the factorization is evaluated first on the host, we use `cusolverRfSetupHost` afterwards (all the sparse matrices and the  permutation matrices are instantied on the host). If we could evaluate the initial factorization on the device, it would be more appropriate to use `cusolverRfSetupDevice` indeed. Unfortunately, `cusolver` does not provide any function to compute such LU factorization directly on the device (as far as I am aware of). 

But I think we might want to change the API to reflect that (ideally the constructor of `RfLU` should take as input a factorization, according to the previous discussion). However, we should be able to extract both the permutation matrices `P` and `Q` and the factorized matrices `L, U` to initiate a new `RfLU` object with `cusolverRfSetup{Host,Device}`, and I am not sure there is a unified API to do so in Julia...